src/detect-engine-alert.c

```diff
@@ -206,14 +200,18 @@ int PacketAlertAppend(DetectEngineThreadCtx *det_ctx, const Signature *s,
         p->alerts.alerts[p->alerts.cnt].frame_id =
                 (flags & PACKET_ALERT_FLAG_FRAME) ? det_ctx->frame_id : 0;
     } else {
-        /* We need to make room for this s->num
-         (a bit ugly with memcpy but we are planning changes here)*/
-        for (i = p->alerts.cnt - 1; i >= 0 && p->alerts.alerts[i].num > s->num; i--) {
-            memcpy(&p->alerts.alerts[i + 1], &p->alerts.alerts[i], sizeof(PacketAlert));
-        }
+        /* find position to insert */
+        for (i = p->alerts.cnt - 1; i >= 0 && p->alerts.alerts[i].num > s->num; i--)
```

Comment:
There is a duplicate test for `p->alerts.alerts[p->alerts.cnt - 1].num < s->num`

Reply:
And 
`if (p->alerts.cnt == 0 || (p->alerts.cnt > 0 &&
                               p->alerts.alerts[p->alerts.cnt - 1].num < s->num)) {`
can be simplified to
`if (p->alerts.cnt == 0 || 
                               p->alerts.alerts[p->alerts.cnt - 1].num < s->num) {`
as alerts.cnt is unsigned cf `PacketAlerts_`

Reply:
> There is a duplicate test for `p->alerts.alerts[p->alerts.cnt - 1].num < s->num`

Sorry, I don't think I get this. Can you point out where?

Reply:
It is tested a first time in the surrounding if. Then in the else clause, it is tested again in the loop

Reply:
I think I see it now. Because the surrounding if has it `<` while the for checks for `>` I didn't realize it before. Thanks :)